lib/optimizely/semantic_version.rb

```diff
@@ -27,14 +27,43 @@ module SemanticVersion
     module_function
 
     def pre_release?(target)
-      # Method to check if the given version contains "-"
+      # Method to check if the given version is a prerelease
       #
       # target - String representing semantic version
       #
-      # Returns true if the given version does contain "-"
+      # Returns true if the given version is a prerelease
       #         false if it doesn't
 
-      target.include? SEMVER_PRE_RELEASE
+      raise unless target.is_a? String
+
+      prerelease_index = target.index(SEMVER_PRE_RELEASE)
+      build_index = target.index(SEMVER_BUILD)
+
+      return false if prerelease_index.nil?
+      return true if build_index.nil?
+
+      # when both operators are present prerelease should precede the build operator
+      prerelease_index < build_index
+    end
+
+    def build?(target)
+      # Method to check if the given version is a build
+      #
+      # target - String representing semantic version
+      #
+      # Returns true if the given version is a build
+      #         false if it doesn't
+
+      raise unless target.is_a? String
+
+      prerelease_index = target.index(SEMVER_PRE_RELEASE)
+      build_index = target.index(SEMVER_BUILD)
+
+      return false if build_index.nil?
+      return true if prerelease_index.nil?
+
+      # when both operators are present build should precede the prerelease operator
+      build_index < prerelease_index
```

Comment:
Have we considered just having a single method called `prerelease_or_build` and that just determines if version is prerelease or build vs doing similar things twice.

Reply:
Calling the index method is not cheap and we can technically halve the number of times we call it.

Reply:
@aliabbasrizvi I would generally agree with you.  However, you need to keep in context your context :D.  Walking a small string like a version string (finding first index is going to be quite small (e.g. 2.0.0 or 2.0.0- or 2.0.0+).  We could always just check the 6th element.  But again, in general you are walking a tiny string.

Reply:
Sure. I accept that it is small. Ok in current form.